Importamos las librerias que vamos a utilizar.

In [9]:
import pandas as pd
import numpy as np
import ast

In [10]:

# Establecer la opción para mostrar todas las columnas
pd.set_option('display.max_columns', None)

# Carga de datos.

In [11]:
movies_raw = pd.read_csv('data/movies_dataset.csv',low_memory=False)

---

# Analisis de movies_raw.

In [12]:
movies_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

Tenemos 45466 filas y 24 columnas. 
Algunas columnas presentan inconsistencias en sus datos y valores nulos.

In [13]:
movies_raw.describe(include='all')

adult                              belongs_to_collection budget  \
count   45466                                               4494  45466   
unique      5                                               1698   1226   
top     False  {'id': 415931, 'name': 'The Bowery Boys', 'pos...      0   
freq    45454                                                 29  36573   
mean      NaN                                                NaN    NaN   
std       NaN                                                NaN    NaN   
min       NaN                                                NaN    NaN   
25%       NaN                                                NaN    NaN   
50%       NaN                                                NaN    NaN   
75%       NaN                                                NaN    NaN   
max       NaN                                                NaN    NaN   

                               genres                     homepage      id  \
count                           45466                         7782   45466   
unique                           4069                         7673   45436   
top     [{'id': 18, 'name': 'Drama'}]  http://www.georgecarlin.com  141971   
freq                             5000                           12       3   
mean                              NaN                          NaN     NaN   
std                               NaN                          NaN     NaN   
min                               NaN                          NaN     NaN   
25%                               NaN                          NaN     NaN   
50%                               NaN                          NaN     NaN   
75%                               NaN                          NaN     NaN   
max                               NaN                          NaN     NaN   

          imdb_id original_language original_title            overview  \
count       45449             45455          45466               44512   
unique      45417                92          43373               44307   
top     tt1180333                en         Hamlet  No overview found.   
freq            3             32269              8                 133   
mean          NaN               NaN            NaN                 NaN   
std           NaN               NaN            NaN                 NaN   
min           NaN               NaN            NaN                 NaN   
25%           NaN               NaN            NaN                 NaN   
50%           NaN               NaN            NaN                 NaN   
75%           NaN               NaN            NaN                 NaN   
max           NaN               NaN            NaN                 NaN   

       popularity                       poster_path production_companies  \
count       45461                             45080                45463   
unique      43758                             45024                22708   
top           0.0  /5D7UBSEgdyONE6Lql6xS7s6OLcW.jpg                   []   
freq           66                                 5                11875   
mean          NaN                               NaN                  NaN   
std           NaN                               NaN                  NaN   
min           NaN                               NaN                  NaN   
25%           NaN                               NaN                  NaN   
50%           NaN                               NaN                  NaN   
75%           NaN                               NaN                  NaN   
max           NaN                               NaN                  NaN   

                                     production_countries release_date  \
count                                               45463        45379   
unique                                               2393        17336   
top     [{'iso_3166_1': 'US', 'name': 'United States o...   2008-01-01   
freq                                                17851          136  

In [14]:
movies_raw.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943   
1  When siblings Judy and Peter discover an encha...  17.015539   
2  A family wedding reignites the ancient feud be...    11.7129   
3  Cheated on, mistreated and stepped on, the wom...   3.859495   
4  Just when George Banks has recovered from his ...   8.387519   

                        poster_path  \
0  /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1  /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2  /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3  /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4  /e64sOI48hQXyru7naBFyssKFxVd.jpg   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3  [{'name': 'Twentieth Century Fox Film Corporat...   
4  [{'name': 'Sandollar Productions', 'id': 5842}...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue  runtime                                   spoken_languages  \
0  373554033.0     81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0    104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0    101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0    127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0    106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video  vote_average  vote_count  
0                    Toy Story  False           7.7      5415.0  
1                      Jumanji  False           6.9      2413.0  
2             Grumpier Old Men  False           6.5        92.0  
3            Waiting to Exh

Podemos observar 5 columnas que contienen datos anidados. Estas son: 'belongs_to_collection', 'genres', 'production_companies', 'production_countries' y 'spoken_languages'.

---

# Valores Duplicados y nulos.

¿Hay filas duplicadas? ¿Cuántas?


In [15]:
#filas duplicadas
(
    movies_raw
    .duplicated(keep=False)
    .sum()
)

33

In [16]:
# Filas con campo id duplicado /  Una misma pelicula aparece varias veces en el dataframe.
(
    movies_raw
    .duplicated(subset='id',keep=False)
    .sum()
)

59

Eliminaremos las filas duplicadas y las que poseen ids duplicados. 
Ya que representan menos del 1% del dataset.
Pero se consultará para posteriores entregas.

¿Cuántos datos nulos tiene cada columna?

In [17]:
(
    movies_raw
    .isnull()
    .sum()
)

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

¿Cuáles son las 6 peliculas que no tienen título especificado?

In [18]:
movies_raw[movies_raw['title'].isnull()]

adult  \
19729                                              False   
19730                                 - Written by Ørnås   
29502                                              False   
29503   Rune Balot goes to a casino connected to the ...   
35586                                              False   
35587   Avalanche Sharks tells the story of a bikini ...   

                                   belongs_to_collection  \
19729                                                NaN   
19730                                           0.065736   
29502  {'id': 122661, 'name': 'Mardock Scramble Colle...   
29503                                           1.931659   
35586                                                NaN   
35587                                           2.185485   

                                 budget  \
19729                                 0   
19730  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29502                                 0   
29503  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35586                                 0   
35587  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19729  [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29502  [{'id': 16, 'name': 'Animation'}, {'id': 878, ...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35586  [{'id': 10770, 'name': 'TV Movie'}, {'id': 28,...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage          id  \
19729                                                NaN       82663   
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20   
29502                      http://m-scramble.jp/exhaust/      122662   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29   
35586                                                NaN      249260   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01   

         imdb_id original_language                            original_title  \
19729  tt0113002                en                              Midnight Man   
19730          0             104.0  [{'iso_639_1': 'en', 'name': 'English'}]   
29502  tt2423504                ja                          マルドゥック・スクランブル 排気   
29503          0              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]   
35586  tt2622826                en                          Avalanche Sharks   
35587          0              82.0  [{'iso_639_1': 'en', 'name': 'English'}]   

                                                overview  \
19729  British soldiers force a recently captured IRA...   
19730                                           Released   
29502         Third film of the Mardock Scramble series.   
29503                                           Released   
35586  A group of skiers are terrorized during spring...   
35587                                           Released   

                  popularity                          poster_path  \
19729                    NaN                                  NaN   
19730                    NaN                         Midnight Man   
29502                    NaN                                  NaN   
29503                    NaN  Mardock Scramble: The Third Exhaust   
35586                    NaN                                  NaN   
35587  Beware Of Frost Bites                     Avalanche Sharks   

      production_companies production_countries release_date  revenue  \
19729                  NaN                  NaN          NaN      NaN   
19730                False                  6.0            1      NaN   
29502                  NaN                  NaN          NaN      NaN   
29503                False                  7.0           12      NaN   
35586                  NaN                  NaN          NaN      NaN   
35587                False                  4.3           22      NaN   

   

Estas 6 filas también se eliminaran ya que la mayoría de sus datos son nulos o inconsistentes.

---

# Creación del dataframe movies.

In [19]:
movies = (
    movies_raw[
        ['id',
        'title',
        'tagline',
        'overview',
        'original_language',
        'runtime',
        'status',
        'release_date',
        # release_year
        # id_collection
        # return
        'budget',
        'revenue',
        'vote_count',
        'vote_average', 
        'popularity',
        'belongs_to_collection',
        'genres',
        'spoken_languages',
        'production_companies',
        'production_countries'
        ]]
)



In [20]:
movies.shape

(45466, 18)

# Datos duplicados y nulos

Eliminamos las filas duplicadas y las que poseen ids duplicados ya que representan menos del 1% de los datos. 
Pero se tendran en cuenta para próximas entregas.

In [21]:
movies = movies.drop_duplicates()

movies = movies.drop_duplicates(subset='id',keep=False)

# movies[movies['id']==str(10991)]  # revisamos si se borraron todas las filas duplicadas con un id de prueba

# # revimamos si quedaron ids duplicados 
# print(movies[
#     movies
#     .duplicated(subset='id',keep=False)   
# ].shape)
# # ids_duplicados.sort_values(by='id')[['id','title']] #para ordenarlos

In [22]:
movies.shape

(45423, 18)

Eliminamos las 6 filas que poseen valores nulos en el campo title. Ya que poseen valores inconsistentes y muchos valores nulos.

In [23]:
movies = movies.drop(movies[movies['title'].isnull()].index)
movies.shape

(45417, 18)

El campo release_date tiene 87 valores nulos que deben eliminarse
Los valores nulos del campo release date deben eliminarse. Debido a que es item de la Propuesta de trabajo.

In [24]:
# movies[movies['release_date'].isnull()].head(5) # filtrar las filas con release_date NAN

In [25]:
#Eliminar los release_date NAN // release_date      
movies = movies.dropna(subset='release_date')

In [26]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45333 entries, 0 to 45465
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     45333 non-null  object 
 1   title                  45333 non-null  object 
 2   tagline                20381 non-null  object 
 3   overview               44392 non-null  object 
 4   original_language      45322 non-null  object 
 5   runtime                45087 non-null  float64
 6   status                 45253 non-null  object 
 7   release_date           45333 non-null  object 
 8   budget                 45333 non-null  object 
 9   revenue                45333 non-null  float64
 10  vote_count             45333 non-null  float64
 11  vote_average           45333 non-null  float64
 12  popularity             45333 non-null  object 
 13  belongs_to_collection  4483 non-null   object 
 14  genres                 45333 non-null  object 
 15  spoken_

# Tipo de datos y creacion de la columna 'release_year'

Cambiamos los tipos de los datos para que estos sean consistentes con los datos que contienen.

In [27]:
#cambiamos los tipos de datos de las columnas
movies = movies.rename({'id': 'id_movie'}, axis=1)

movies['title'] = movies['title'].apply(lambda x: x.title())

movies = movies.astype({'id_movie' : 'int64', 
                        # 'title':'str',
                        # 'tagline':'str',
                        # 'overview':'str',
                        'original_language':'category',
                        # # 'runtime':'int', # posee valores nan por lo que aun no podemos cambiarlo a int
                        'status': 'category', 
                        # 'release_date': ,  # no se si se puede hacer aca directamente
                        'budget': 'int64', 
                        'revenue': 'int64', # posee valores nan por lo que aun no podemos cambiarlo a int
                        'vote_count': 'int64', # posee valores nan por lo que aun no podemos cambiarlo a int
                        'vote_average': 'int', # posee valores nan por lo que aun no podemos cambiarlo a int
                        # 'popularity' : 'float' # no se que es esto jajaj 
                        }, copy=False, errors='raise')

De haber fechas, deberán tener el formato AAAA-mm-dd.

In [28]:
#cambiamos los tipos de datos de la columna 'release_date'
movies['release_date'] = pd.to_datetime(movies['release_date'], format='%Y-%m-%d',yearfirst=True, errors='coerce')

A demás deberán crear la columna release_year donde extraerán el año de la fecha de estreno

In [29]:
#Creamos la columna 'release_year'
movies.insert(movies.columns.get_loc('release_date') + 1, 'release_year', movies['release_date'].dt.year)

In [30]:
movies.head(3)

id_movie             title  \
0       862         Toy Story   
1      8844           Jumanji   
2     15602  Grumpier Old Men   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   

                                            overview original_language  \
0  Led by Woody, Andy's toys live happily in his ...                en   
1  When siblings Judy and Peter discover an encha...                en   
2  A family wedding reignites the ancient feud be...                en   

   runtime    status release_date  release_year    budget    revenue  \
0     81.0  Released   1995-10-30          1995  30000000  373554033   
1    104.0  Released   1995-12-15          1995  65000000  262797249   
2    101.0  Released   1995-12-22          1995         0          0   

   vote_count  vote_average popularity  \
0        5415             7  21.946943   
1        2413             6  17.015539   
2          92             6    11.7129   

                               belongs_to_collection  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                NaN   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                                    spoken_languages  \
0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2           [{'iso_639_1': 'en', 'name': 'English'}]   

                                production_companies  \
0     [{'name': 'Pixar Animation Studios', 'id': 3}]   
1  [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2  [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   

                                production_countries  
0  [{'iso_3166_1': 'US', 'name': 'United States o...  
1  [{'iso_3166_1': 'US', 'name': 'United States o...  
2  [{'iso_3166_1': 'US', 'name': 'United States o...

---

Los valores nulos de los campos revenue, budget deben ser rellenados por el número 0.

In [31]:
# Los valores nulos de los campos revenue, budget deben ser rellenados por el número 0
movies['budget'].fillna(0, inplace=True)
movies['revenue'].fillna(0, inplace=True)


---


Crear la columna con el retorno de inversión, llamada return con los campos revenue y budget, dividiendo estas dos últimas revenue / budget, cuando no hay datos disponibles para calcularlo, deberá tomar el valor 0.

In [32]:
# funcion que permite calcular el return
def ret(df,col1,col2):
    if df[col1] == 0:
        out = 0

        
    else:    
        out = df[col2] / df[col1]
    return out

# creamos la columna return
movies.insert(movies.columns.get_loc('revenue') + 1, 'return', movies.apply(ret,axis=1,args={'budget','revenue'}))

----

# Creación de tablas a partir de las columnas con datos anidados

----------

# Creación de la tabla 'release_date_table'

In [33]:
#TABLA LANZAMIENTOS

release_date_table = movies[['id_movie','release_date']].copy()

release_date_table.loc['year'] = movies['release_date'].dt.year

release_date_table['month'] = movies['release_date'].dt.month_name(locale='es_ES.UTF-8').str.lower()

release_date_table['day_num'] = movies['release_date'].dt.day

release_date_table['day_name'] = movies['release_date'].dt.day_name(locale='es_ES.UTF-8').str.lower()
release_date_table['day_name'] = release_date_table['day_name'].replace({'miã©rcoles': 'miércoles', 'sã¡bado': 'sábado'})


In [34]:
release_date_table

id_movie release_date      month  day_num day_name
0         862.0   1995-10-30    octubre     30.0    lunes
1        8844.0   1995-12-15  diciembre     15.0  viernes
2       15602.0   1995-12-22  diciembre     22.0  viernes
3       31357.0   1995-12-22  diciembre     22.0  viernes
4       11862.0   1995-02-10    febrero     10.0  viernes
...         ...          ...        ...      ...      ...
45462  111109.0   2011-11-17  noviembre     17.0   jueves
45463   67758.0   2003-08-01     agosto      1.0  viernes
45464  227506.0   1917-10-21    octubre     21.0  domingo
45465  461257.0   2017-06-09      junio      9.0  viernes
year        NaN          NaT        NaN      NaN      NaN

[45334 rows x 5 columns]

In [35]:
(
    release_date_table
    .isnull()
    .sum()
)

id_movie        1
release_date    1
month           1
day_num         1
day_name        1
dtype: int64

In [36]:
release_date_table = release_date_table.dropna()

In [37]:
#crear el csv 
release_date_table.to_csv("data/release_date.csv",index=False)

Creacion de las tablas estrenos_por_dia y estrenos_por_mes que se utilizaran para las consultas.

In [38]:
estrenos_por_dia = release_date_table['day_name'].value_counts().reset_index()
estrenos_por_dia.columns = ['dia', 'cantidad']
estrenos_por_dia

estrenos_por_mes = release_date_table['month'].value_counts().reset_index()
estrenos_por_mes.columns = ['mes', 'cantidad']
estrenos_por_mes

#crear los csv 
estrenos_por_dia.to_csv("data/estrenos_por_dia.csv",index=False)
estrenos_por_mes.to_csv("data/estrenos_por_mes.csv",index=False)


---

# Creación de la tabla 'collections'

In [39]:
# TABLA COLLECTIONS

# Extrae los datos anidados en la columna "belongs_to_collection"
collection = pd.json_normalize(data=movies['belongs_to_collection'].dropna().apply(eval))

collection = pd.DataFrame(collection)

collection = collection[['id','name']]
collection = collection.rename({'id':'id_collections'},axis=1)

#crear el csv 
collection.to_csv("data/collection.csv",index=False)

# belongs_to_collection.head()
# belongs_to_collection.shape

In [40]:
# cremaos la columna 'id_collecction', extraemos sus valores de la columna belongs_to_collection y luego eliminamos esta última.

# # Definimos una expresión regular para extraer el id del campo "id" de la columna belongs_to_collection
regex = r"'id':\s*([\d]+)"
'''
r": esto indica que la cadena debe ser interpretada como una "cadena sin procesar" en Python, lo que significa que cualquier carácter especial se interpreta literalmente, sin ser escapado.
'id': esto es una cadena literal que busca el texto exacto 'id'.
:\s*: esto busca un carácter de dos puntos seguido de cero o más espacios en blanco. La secuencia \s* busca cualquier número de caracteres de espacio en blanco (espacios, tabuladores, saltos de línea, etc.) que puedan aparecer después del carácter de dos puntos.
([\d]+): esto busca uno o más caracteres numéricos, que representan el valor de id. La expresión [\d]+ indica que se busca uno o más caracteres que sean dígitos numéricos (\d es una abreviatura para "dígitos numéricos").
En resumen, la expresión regular busca la cadena 'id': seguida de cualquier número de espacios en blanco, y luego busca uno o más dígitos numéricos para capturar el valor de id. Cuando se aplica esta expresión regular a una cadena JSON que contiene un valor de id, se captura y extrae el valor de id.
'''

# # Utilizamos la función "str.extract" para extraer el id y guardarlo en una nueva columna 'id_collection'
movies['id_collection'] = movies['belongs_to_collection'].str.extract(regex)

# # moverla al principio para comparar
# movies = movies.insert(0, 'id_collection', movies.pop('id_collection'))

# Eliminamos la columna 'belongs_to_collection'
movies.drop(['belongs_to_collection'], axis= 'columns', inplace=True)

# movies.head(2)

--------

# Creación de la tabla 'movies_production_companies'

In [41]:
### TABLA MOVIES_production_companies
movies_production_companies = movies[['id_movie','production_companies']].copy()

#reemplazamos los valores nulos por "[{'name': 'None', 'id': 0}]"
movies_production_companies['production_companies'] = movies_production_companies[['production_companies']].replace("[]", "[{'name': 'None', 'id': 0}]")

movies_production_companies['production_companies'] = movies_production_companies['production_companies'].apply(ast.literal_eval)

# Aplicamos explode() para crear filas adicionales para cada elemento en la lista production_companies
movies_production_companies = movies_production_companies.explode('production_companies')

# Extraemos el valor de id_production_companies de cada elemento en la lista production_companies
movies_production_companies['id_production_companies'] = movies_production_companies['production_companies'].apply(lambda x: x['id'])

# # Eliminar la columna production_companies
movies_production_companies = movies_production_companies.drop('production_companies', axis=1)

# Creamos el csv
movies_production_companies.to_csv("data/movies_production_companies.csv",index=False)


# TABLA production_companies

production_companies = movies['production_companies'].apply(ast.literal_eval) # Convertir el string a lista de diccionarios
production_companies = pd.DataFrame([dict(x) for x in production_companies.sum()])[['id', 'name']]
production_companies = production_companies.drop_duplicates()
production_companies.to_csv("data/production_companies.csv",index=False)


# Eliminamos la columna production_companies de la tabla movies
movies = movies.drop('production_companies', axis=1)

-------

# Creación de la tabla 'movies_production_countries'

In [42]:
### TABLA MOVIES_production_countries

movies_production_countries = movies[['id_movie','production_countries']].copy()

# Reemplazamos los valores nulos por "[{'iso_3166_1': 'none', 'name': 'none'}]"
movies_production_countries['production_countries'] = movies_production_countries[['production_countries']].replace("[]", "[{'iso_3166_1': 'none', 'name': 'none'}]")  # ya esta arriba
movies_production_countries['production_countries'] = movies_production_countries['production_countries'].apply(ast.literal_eval)

# Aplicamos explode() para crear filas adicionales para cada elemento en la lista production_countries
movies_production_countries = movies_production_countries.explode('production_countries')

# Extraer el valor de id_production_countries de cada elemento en la lista production_countries
movies_production_countries['id_production_countries'] = movies_production_countries['production_countries'].apply(lambda x: x['iso_3166_1'])

# Eliminar la columna genres
movies_production_countries = movies_production_countries.drop('production_countries', axis=1)

# Creamos el csv
movies_production_countries.to_csv("data/movies_production_countries.csv",index=False)


### TABLA production_countries

production_countries = movies['production_countries'].apply(ast.literal_eval) # Convertir el string a lista de diccionarios
production_countries = pd.DataFrame([dict(x) for x in production_countries.sum()])[['iso_3166_1', 'name']]
production_countries = production_countries.drop_duplicates()

# Creamos el csv
production_countries.to_csv("data/production_countries.csv",index=False)

# Eliminamos la columna production_countries de la tabla movies
movies = movies.drop('production_countries', axis=1)


-------

# Creación de la tabla 'movies_spoken_languages'

In [43]:
### TABLA MOVIES_spoken_languages

movies_spoken_languages = movies[['id_movie','spoken_languages']].copy()

# Reemplazamos los valores nulos por "[{'iso_639_1': 'none', 'name': 'none'}]"
movies_spoken_languages['spoken_languages'] = movies_spoken_languages[['spoken_languages']].replace("[]", "[{'iso_639_1': 'none', 'name': 'none'}]")  
movies_spoken_languages['spoken_languages'] = movies_spoken_languages['spoken_languages'].apply(ast.literal_eval)

# Aplicamos explode() para crear filas adicionales para cada elemento en la lista genres
movies_spoken_languages = movies_spoken_languages.explode('spoken_languages')

# Extraer el valor de id_genre de cada elemento en la lista genres
movies_spoken_languages['id_spoken_languages'] = movies_spoken_languages['spoken_languages'].apply(lambda x: x['iso_639_1'])

# Eliminar la columna spoken_languages
movies_spoken_languages = movies_spoken_languages.drop('spoken_languages', axis=1)

# Creamos el csv
movies_spoken_languages.to_csv("data/movies_spoken_languages.csv",index=False)


### TABLA spoken_languages

spoken_languages = movies['spoken_languages'].apply(ast.literal_eval) # Convertir el string a lista de diccionarios
spoken_languages = pd.DataFrame([dict(x) for x in spoken_languages.sum()])[['iso_639_1', 'name']]
spoken_languages = spoken_languages.drop_duplicates()

# Creamos el csv
spoken_languages.to_csv("data/spoken_languages.csv",index=False)


# Eliminamos la columna production_countries de la tabla movies
movies = movies.drop('spoken_languages', axis=1)

-------

# Creación de la tabla 'genres'

In [44]:
# TABLA GENRES

genres = movies['genres'].apply(ast.literal_eval) # Convertir el string a lista de diccionarios
genres = pd.DataFrame([dict(x) for x in genres.sum()])[['id', 'name']]
genres = genres.drop_duplicates()
genres.to_csv("data/genres.csv",index=False)


### TABLA MOVIES_GENRES
movies_genres = movies[['id_movie','genres']].copy()

# Generamos la categoria none para las peliculas que no tienen un genero especificado
movies_genres['genres'] = movies_genres[['genres']].replace("[]", "[{'id': 0, 'name': 'None'}]")  # ya esta arriba

movies_genres['genres'] = movies_genres['genres'].apply(ast.literal_eval)

# # Aplicamos explode() para crear filas adicionales para cada elemento en la lista genres
movies_genres = movies_genres.explode('genres')

# # Extraer el valor de id_genre de cada elemento en la lista genres
movies_genres['id_genre'] = movies_genres['genres'].apply(lambda x: x['id'])

# Eliminar la columna genres
movies_genres = movies_genres.drop('genres', axis=1)


# 
movies = movies.drop('genres', axis=1)


movies_genres.to_csv("data/movies_genres.csv",index=False)



In [45]:
movies.head()

id_movie                        title  \
0       862                    Toy Story   
1      8844                      Jumanji   
2     15602             Grumpier Old Men   
3     31357            Waiting To Exhale   
4     11862  Father Of The Bride Part Ii   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                                            overview original_language  \
0  Led by Woody, Andy's toys live happily in his ...                en   
1  When siblings Judy and Peter discover an encha...                en   
2  A family wedding reignites the ancient feud be...                en   
3  Cheated on, mistreated and stepped on, the wom...                en   
4  Just when George Banks has recovered from his ...                en   

   runtime    status release_date  release_year    budget    revenue  \
0     81.0  Released   1995-10-30          1995  30000000  373554033   
1    104.0  Released   1995-12-15          1995  65000000  262797249   
2    101.0  Released   1995-12-22          1995         0          0   
3    127.0  Released   1995-12-22          1995  16000000   81452156   
4    106.0  Released   1995-02-10          1995         0   76578911   

     return  vote_count  vote_average popularity id_collection  
0  0.080310        5415             7  21.946943         10194  
1  0.247339        2413             6  17.015539           NaN  
2  0.000000          92             6    11.7129        119050  
3  0.196434          34             6   3.859495           NaN  
4  0.000000         173             5   8.387519         96871

In [46]:
movies.to_csv("data/movies_etl.csv",index=False)
# .


In [47]:
movies

id_movie                        title  \
0           862                    Toy Story   
1          8844                      Jumanji   
2         15602             Grumpier Old Men   
3         31357            Waiting To Exhale   
4         11862  Father Of The Bride Part Ii   
...         ...                          ...   
45460     30840                   Robin Hood   
45462    111109          Century Of Birthing   
45463     67758                     Betrayal   
45464    227506             Satan Triumphant   
45465    461257                     Queerama   

                                                 tagline  \
0                                                    NaN   
1              Roll the dice and unleash the excitement!   
2      Still Yelling. Still Fighting. Still Ready for...   
3      Friends are the people who let you be yourself...   
4      Just When His World Is Back To Normal... He's ...   
...                                                  ...   
45460                                                NaN   
45462                                                NaN   
45463                             A deadly game of wits.   
45464                                                NaN   
45465                                                NaN   

                                                overview original_language  \
0      Led by Woody, Andy's toys live happily in his ...                en   
1      When siblings Judy and Peter discover an encha...                en   
2      A family wedding reignites the ancient feud be...                en   
3      Cheated on, mistreated and stepped on, the wom...                en   
4      Just when George Banks has recovered from his ...                en   
...                                                  ...               ...   
45460  Yet another version of the classic epic, with ...                en   
45462  An artist struggles to finish his work while a...                tl   
45463  When one of her hits goes wrong, a professiona...                en   
45464  In a small town live two brothers, one a minis...                en   
45465  50 years after decriminalisation of homosexual...                en   

       runtime    status release_date  release_year    budget    revenue  \
0         81.0  Released   1995-10-30          1995  30000000  373554033   
1        104.0  Released   1995-12-15          1995  65000000  262797249   
2        101.0  Released   1995-12-22          1995         0          0   
3        127.0  Released   1995-12-22          1995  16000000   81452156   
4        106.0  Released   1995-02-10          1995         0   76578911   
...        ...       ...          ...           ...       ...        ...   
45460    104.0  Released   1991-05-13          1991         0          0   
45462    360.0  Released   2011-11-17          2011         0          0   
45463     90.0  Released   2003-08-01          2003         0          0   
45464     87.0  Released   1917-10-21          1917         0          0   
45465     75.0  Released   2017-06-09          2017         0          0   

         return  vote_count  vote_average popularity id_collection  
0      0.080310        5415             7  21.946943         10194  
1      0.247339        2413             6  17.015539           NaN  
2      0.000000          92             6    11.7129        119050  
3      0.196434          34             6   3.859495           NaN  
4      0.000000         173             5   8.387519         96871  
...         ...         ...           ...        ...           ...  
45460  0.000000          26             5   5.683753           NaN  
45462  0.000000           3             9   0.178241           NaN  
45463  0.000000           6             3   0.903007           NaN  
45464  0.000000           0             0   0.003503           NaN  
45465  0.000000           0             0   0.163015           NaN  

[45333 rows x 16 columns]

--------------